File: 42892.swift
import Foundation

class BinaryTree {
    class Node {
        let number: Int
        let x: Int
        let y: Int
        var leftChild: Node?
        var rightChild: Node?

        init(number: Int, x: Int, y: Int) {
            self.number = number
            self.x = x
            self.y = y
        }
    }
    
    var root: Node?
    
    init(nodeInfo: [[Int]]) {
        var nodes = nodeInfo.enumerated().map { Node(number: $0 + 1, x: $1[0], y: $1[1]) }.sorted { $0.y < $1.y || ($0.y == $1.y && $0.x > $1.x) }
        while let current = nodes.popLast() { self.appendNew(current, self.root) }
    }
    
    private func appendNew(_ node: Node, _ parent: Node? = nil) {
        guard let parentNode = parent else { return self.root = node }
        if node.x < parentNode.x {
            guard let leftChild = parentNode.leftChild else { return parentNode.leftChild = node }
            appendNew(node, leftChild)
        } else {
            guard let rightChild = parentNode.rightChild else { return parentNode.rightChild = node }
            appendNew(node, rightChild)
        }
    }
    
    func traversePreorder(result: inout [Int]) {
        self.traversePreorderRecursively(node: self.root, result: &result)
    }
    
    func traversePostorder(result: inout [Int]) {
        self.traversePostorderRecursively(node: self.root, result: &result)
    }

    private func traversePreorderRecursively(node: Node?, result: inout [Int]) {
        if let current = node {
            result.append(current.number)
            traversePreorderRecursively(node: current.leftChild, result: &result)
            traversePreorderRecursively(node: current.rightChild, result: &result)
        }
    }
    
    private func traversePostorderRecursively(node: Node?, result: inout [Int]) {
        if let current = node {
            traversePostorderRecursively(node: current.leftChild, result: &result)
            traversePostorderRecursively(node: current.rightChild, result: &result)
            result.append(current.number)
        }
    }
}

func solution(_ nodeinfo:[[Int]]) -> [[Int]] {
    let tree = BinaryTree(nodeInfo: nodeinfo)
    var preorderResult: [Int] = []
    var postorderResult: [Int] = []
    
    tree.traversePreorder(result: &preorderResult)
    tree.traversePostorder(result: &postorderResult)
    
    return [preorderResult, postorderResult]
}

File: 42892.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

struct Node {
    int number, x, y;
    Node *leftChild, *rightChild;
};

bool compareNodes(Node left, Node right) {
    return (left.y > right.y || (left.y == right.y && left.x < right.x));
}

void addChild(Node *current, Node *parent) {
    if(parent == NULL) { return; }
    
    if(current->x < parent->x) {
        if(parent->leftChild == NULL) { parent->leftChild = current; }
        else { addChild(current, parent->leftChild); }
    } else {
        if(parent->rightChild == NULL) { parent->rightChild = current; }
        else { addChild(current, parent->rightChild); }
    }
}

void preorderTraverse(vector<int>* result, Node* current) {
    if(current == NULL) { return; }
    result->push_back(current->number);
    preorderTraverse(result, current->leftChild);
    preorderTraverse(result, current->rightChild);
}

void postorderTraverse(vector<int>* result, Node* current) {
    if(current == NULL) { return; }
    postorderTraverse(result, current->leftChild);
    postorderTraverse(result, current->rightChild);
    result->push_back(current->number);
}

vector<vector<int>> solution(vector<vector<int>> nodeinfo) {
    vector<int> preorderTraversalResult;
    vector<int> postorderTraversalResult;
    vector<Node> nodes;
    Node *root;
    
    for(int i = 0; i < nodeinfo.size(); i++) {
        Node node;
        node.number = i + 1;
        node.x = nodeinfo[i][0];
        node.y = nodeinfo[i][1];
        node.leftChild = NULL;
        node.rightChild = NULL;
        nodes.push_back(node);
    }
    
    sort(nodes.begin(), nodes.end(), compareNodes);
    root = &nodes[0];
    
    for(int i = 1; i < nodes.size(); i++) { addChild(&nodes[i], root); }
    preorderTraverse(&preorderTraversalResult, root);
    postorderTraverse(&postorderTraversalResult, root);
    
    return vector<vector<int>>{preorderTraversalResult, postorderTraversalResult};
}
